cpg-language-python/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/python/StatementHandler.kt

```diff
@@ -394,95 +395,102 @@ class StatementHandler(frontend: PythonLanguageFrontend) :
-                newProblemDeclaration(
-                    "`kwonlyargs` are not yet supported",
-                    problemType = ProblemNode.ProblemType.TRANSLATION,
-                    rawNode = args
-                )
-            frontend.scopeManager.addDeclaration(problem)
-        }
+        handleKeywordOnlyArguments(args)
+    }
 
-        if (args.kw_defaults.isNotEmpty()) {
+    private fun handleReceiverArgument(
+        positionalArguments: List<Python.ASTarg>,
+        args: Python.ASTarguments,
+        result: FunctionDeclaration,
+        recordDeclaration: RecordDeclaration
+    ) {
+        // first argument is the receiver
+        val recvPythonNode = positionalArguments.firstOrNull()
+        if (recvPythonNode == null) {
             val problem =
                 newProblemDeclaration(
-                    "`kw_defaults` are not yet supported",
+                    "Expected a receiver",
                     problemType = ProblemNode.ProblemType.TRANSLATION,
                     rawNode = args
                 )
             frontend.scopeManager.addDeclaration(problem)
+        } else {
+            val tpe = recordDeclaration.toType()
+            val defaultValue =
+                args.defaults.getOrNull(0)?.let { frontend.expressionHandler.handle(it) }
+            val recvNode =
+                newVariableDeclaration(
+                    name = recvPythonNode.arg,
+                    type = tpe,
+                    implicitInitializerAllowed = false,
+                    rawNode = recvPythonNode
+                )
+            if (defaultValue != null) {
+                val problem =
+                    newProblemDeclaration(
+                        "Receiver with default value",
+                        problemType = ProblemNode.ProblemType.TRANSLATION,
+                        rawNode = args
+                    )
+                frontend.scopeManager.addDeclaration(problem)
+            }
+            frontend.scopeManager.addDeclaration(recvNode)
+            when (result) {
+                is ConstructorDeclaration -> result.receiver = recvNode
+                is MethodDeclaration -> result.receiver = recvNode
+                else -> TODO()
+            }
         }
+    }
 
-        args.kwarg?.let {
-            val problem =
-                newProblemDeclaration(
-                    "`kwarg` is not yet supported",
-                    problemType = ProblemNode.ProblemType.TRANSLATION,
-                    rawNode = it
-                )
-            frontend.scopeManager.addDeclaration(problem)
+    private fun handlePositionalArguments(
+        positionalArguments: List<Python.ASTarg>,
+        args: Python.ASTarguments,
+        startIdx: Int
+    ) {
+        val nonDefaultArgsCount = positionalArguments.size - args.defaults.size
+
+        for ((idx, arg) in positionalArguments.drop(startIdx).withIndex()) {
+            val defaultIndex = idx - nonDefaultArgsCount
```

Comment:
This indexing math looks off to me. Assume `foo(self, a, b = 2, c = 3):` then
- `nonDefaultArgsCount = 4 - 2 = 2`
- `positionalArguments.drop(startIdx)` => this will drop the receiver => a list of 3 elements with indices 0,1,2
- `defaultIndex = idx - nonDefaultArgsCount = 0 - 2 = -2` ???

Either I'm confused or this is off.

Reply:
I just tried this with the debugger -> there is an error as this code does not find the default for `b`.
Please also include a test case for:
```Python
class Foo:
  def foo(self, a, b = 1, c = 2):
    pass
```

Reply:
The indexing math ensures that default arguments are applied correctly by first dropping the receiver. The defaultIndex is calculated by subtracting the number of non-default arguments, which might result in negative values for arguments without a default. However, this is intentional because we only use defaultIndex when it's greater than or equal to 0, meaning it only applies to arguments with defaults.

I added another test with self, which like the others also passes.

Reply:
Please see this test https://github.com/Fraunhofer-AISEC/cpg/pull/1669/commits/8b15388ed69913a33abbe6e48d37efe5b09cecc2 (feel free to edit / update it as I currently do not have access to a Java IDE).

Reply:
Also: it would be great to have your answer in the code as to get a better understanding what and why things are happening here. Maybe a URL to the Python docs where the argument handling is introduced?

Reply:
I refactored the method and also added a doc string. Thank you for adding an additional test. Now this test also passes.